gateway-ha/src/main/java/io/trino/gateway/ha/security/util/BasicCredentials.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.gateway.ha.security.util;
+
+import com.google.common.base.Splitter;
+import jakarta.ws.rs.container.ContainerRequestContext;
+
+import java.util.Base64;
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.base.Strings.nullToEmpty;
+import static com.google.common.net.HttpHeaders.AUTHORIZATION;
+import static java.nio.charset.StandardCharsets.ISO_8859_1;
+import static java.util.Objects.requireNonNull;
+
+public record BasicCredentials(String username, String password)
```

Comment:
Would it make sense to use `io.trino.server.security.BasicAuthCredentials` here?

Reply:
I tried that before. There were few problems:

Some Trino authentication logic are mixed into BasicAuthCredentials. The username must not be empty, otherwise it will throw ([BasicAuthCredentials.java#L61-L63](https://github.com/trinodb/trino/blob/f769e9d3c35d319d16dfe8bb8d2f9f51b4223ca2/core/trino-main/src/main/java/io/trino/server/security/BasicAuthCredentials.java#L61-L63)). That's not required by RFC 2617/7617.

It mixed AuthenticationException and Optional.Empty() to indicate different parsing errors, in order to satisfy different use cases in Trino. That's not needed by gateway and makes the caller's logic become complicated.

It doesn't have unit test coverage in Trino. I don't feel safe to extract and reuse that particular class.